Smt hooks

h0nzZik opened the pull request.

h0nzZik committed:
hook-smt-symbol

BTW apparently, matching hooked declarations in the declarations cell is
slow

h0nzZik committed:
automatically import prelude

h0nzZik committed:
trying to pass a path to the prover (without much success)

h0nzZik committed:
imports system "prelude.kore"

h0nzZik committed:
prover-kore - a wrapper script

h0nzZik committed:
SMT hooked sorts

nishantjr commented on prover/build:
```diff
@@ -100,7 +103,7 @@ def known_passing(test):
     commandline = "'--default-strategy %s'" % strategy_for_test(test)
     return proj.source(test) \
-               .then(rule.variable('flags', '-cCOMMANDLINE=%s' % commandline))
+               .then(rule.variable('flags', ('-cCOMMANDLINE=%s' % commandline) + ' ' + proverPathFlag()))
```
Lets not mix `%` based formatting with string concatenation. Can we add this flag to the `CommandLine` grammar instead of adding a new parameter. That way we would get better error messages when this required flag is not passed.

h0nzZik replied:
We can; just give me an example of how should that parameter look when we call `krun`.

nishantjr replied:
HMm, actually this should be changed to use the `./prover` script instead. I'll add a commit that does that later today

nishantjr commented on prover/drivers/base.md:
```diff
@@ -38,6 +43,8 @@ module PROVER-CONFIGURATION
         <declarations>
           <declaration multiplicity="*" type="Set">  .K </declaration>
         </declarations>
+        <hooked-smt-sorts> .Map </hooked-smt-sorts>
```
Shouldn't these just be axioms? Why do we need a new cell?

h0nzZik replied:
Theoretically, we do not need a new cell and can keep `hook-smt-sort` and `hook-smt-symbol` in the `<declarations>` cell as a declaration, but it slows down the prover significantly (smoke-tests take 15 minutes instead of 3).

nishantjr replied:
I wonder if this is an issue with the ocaml backend. Do the smoke tests really finish in 3 minutes on your machine?! It takes more than 30 on mine.

h0nzZik replied:
It probably was an issue with the ocaml backend; under llvm backend, the times are rougly equal. (something over 1.5 minutes on 16-core machine (more precisely, 8-core + hyperthreading)).

nishantjr commented on prover/drivers/kore-driver.md:
````diff
@@ -57,6 +70,15 @@ Add various standard Kore declarations to the configuration directly:
 ```
 
+For hook-smt-symbol, hook-smt-sort
+```k
````
New line missing between line 73 and 74

nishantjr commented on prover/drivers/kore-driver.md:
````diff
@@ -57,6 +70,15 @@ Add various standard Kore declarations to the configuration directly:
+For hook-smt-symbol, hook-smt-sort
+```k
+  rule <k> hook-smt-symbol(From, To) => .K ...</k>
+       <hooked-smt-symbols>... (.Map => (From |-> To)) </hooked-smt-symbols>
````
Lets use axioms here instead of introducing new cells.

h0nzZik replied:
As above.

nishantjr commented on prover/lang/kore-lang.md:
```diff
@@ -181,13 +181,24 @@ only in this scenario*.
+  syntax SMTLIB2SimpleSymbol
+
+  syntax HookDeclaration ::= "hook-smt-sort" "(" Sort ","  SMTLIB2Sort ")"
+                           | "hook-smt-symbol" "(" Symbol "," SMTLIB2SimpleSymbol ")"
```
Use `syntax Pattern ::= hoot-smt-sort(...)`

so we can do this instead:

```
axiom hook-smt-symbol(store, store)
```

h0nzZik replied:
I would prefer not to make it a pattern, since it cannot be desugared to normal ML, and we cannot use the hook in other patterns. If you insist on that syntax, I would rather make a new nonterminal `Axiom` that is either `Pattern` or `HookDeclaration`.

nishantjr commented on prover/prover-kore:
```diff
@@ -0,0 +1,10 @@
+
+file="$1"
+
+./prover run --definition prover-kore "$file" -cCOMMANDLINE=.CommandLine -cPROVERDIR="\"`pwd`\""
```
This is broken if run from another direwctory. Use add "prove_dir=$(cd $(dirname $0)/..; pwd -P)" to get the prover dir.
It is convention to do this at the first line of the file in case there are any change dirs later.

h0nzZik replied:
The command you provided gives us the root of the repository, not the directory of the prover. Why is there `/..`?

nishantjr reviewed (changes requested):
@h0nzZik The final outcome of this PR should be there are no uses on `lt` `gt` etc in e.g. `getReturnSort` etc.

h0nzZik commented:
@nishantjr I would prefer to merge it right after addressing your comments and rebasing on top of `master`, since (1) it is already reviewed and (2) we want PRs to be small. So I will create a new PR with symbol declarations in the prelude.

h0nzZik committed:
hooks are now in <declarations> cell

h0nzZik committed:
fix prover-kore

h0nzZik committed:
`axiom hook-*` syntax
